File: graph/slick.py
"https://www.spoj.com/problems/UCV2013H/"
from queue import Queue

dx = [0,0,1,-1]
dy = [1,-1,0,0]


while(1):
	n,m = map(int, input().split())
	if n == 0 and m == 0:
		break
	#Read data
	cnt = 0
	slick = 0
	q = Queue()
	V = []
	S = []
	R = {}
	for i in range(0, n):
		S.append(list(map(int, input().split())))
		V.append([False for _ in range(m)])
	#Find all slicks
	for i in range(n):
		for j in range(m):
			slick = 0
			if S[i][j] == 1 and V[i][j] == False:
				#Using BFS
				q.put([i,j])
				slick +=1
				V[i][j] = True
				while(not q.empty()):
					v = q.get()
					# print(V)
					for k in range(4):
						kx = v[0] + dx[k]
						ky = v[1] + dy[k]
						if(kx>=0 and kx < n) and(ky >=0 and ky < m):
						    if V[kx][ky] == False and S[kx][ky] == 1:
						    	q.put([kx,ky])
						    	V[kx][ky] = True
						    	slick+=1
				if R.get(slick) != None:
					R[slick]+=1
				else:
					R[slick] =1
				cnt+=1
	slicks = sorted(R.items(), key=lambda x: x[0])
	print(cnt)
	for i in range(len(slicks)):
		print(slicks[i][0], slicks[i][1])	

File: graph/slick.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <stdio.h>
#define MAX 250
using namespace std;

int main()
{
	int dx[4] = {0,0,1,-1};
	int dy[4] = {1,-1,0,0};
	int n, m;
	map<int,int>slicks;
	int mat[MAX][MAX] = {{0,0}};
	int v[MAX][MAX] = {{0,0}};
	int cnt, slick;
	queue<pair<int,int> >q;
	while(1)
	{
		cin>>n>>m;
		if(n == 0 && m == 0)
		{
			break;
		}
		for(int i = 0; i < n; i++)
		{
			for(int j = 0 ; j < m; j++)
			{
				cin>>mat[i][j];
				v[i][j] = 0;
			}
		}
		slicks.clear();
		cnt = 0;
		//Find slick
		for(int i = 0; i < n; i++)
		{
			for(int j = 0; j < m; j++)
			{
				slick = 0;
				if(mat[i][j] && !v[i][j])
				{
					q.push(make_pair(i,j));
					v[i][j] = 1;
					slick+=1;
					while(!q.empty())
					{
						pair<int,int> point = q.front();
						q.pop();
						for(int k = 0; k < 4; k++)
						{
							int kx = point.first + dx[k];
							int ky = point.second + dy[k];
							if ((kx>=0 && kx < n) && (ky>=0 && ky < m) && (mat[kx][ky]) && (!v[kx][ky]))
							{
								q.push(make_pair(kx,ky));
								v[kx][ky] = 1;
								slick +=1;
							}
						}
					}//end while
					slicks[slick]++;
					cnt+=1;
				}//end if

			}//end for rows
		}//end for cols
		cout<<cnt<<"\n";
		for(auto i = slicks.begin(); i !=slicks.end(); ++i)
		{
			cout<<i->first<<" "<<i->second<<"\n";
		}

	}
	return 0;
}